docs: add user-guide explaining CMO federation (#286)
rhobs/observability-operator: docs/user-guides/federation.md

roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-monitoring-view
subjects:
- kind: ServiceAccount
  name: federate-cmo-ms-prometheus  # 👈  ServiceAccount used in the prometheus deployed by ObO, the SA name is <monitoring stack name>-prometheus.
  namespace: federate-cmo
```

### Create ServiceMonitor for Federation

```yaml
apiVersion: monitoring.rhobs/v1
kind: ServiceMonitor
metadata:
  name: federate-cmo-smon
  namespace: federate-cmo
  labels:
    kubernetes.io/part-of: federate-cmo-ms
    monitoring.rhobs/stack: federate-cmo-ms

spec:
  selector: # 👈 use the prometheus service to create a "dummy" target.
    matchLabels:
      app.kubernetes.io/managed-by: observability-operator
      app.kubernetes.io/name: federate-cmo-ms-prometheus

  endpoints:
  - params:
      'match[]': # 👈 scrape only required metrics from in-cluster prometheus
        - '{__name__=~"container_cpu_.*", namespace="federate-cmo"}'
        - '{__name__="container_memory_working_set_bytes", namespace="federate-cmo"}'
        # NOTE: add additional metrics
        # - '{__name__=~"container_cpu.*|container_memory_.*", namespace="kube-system"}'

>     # relabel example

Author: Lets remove it 
```suggestion
```

# Federate Monitoring Stack and Openshift In-Cluster Prometheus

## Architecture / Topology

![Architecture](federation/assets/cmo-obo-federation.png)

> This example deploy a MonitoringStack in `federate-cmo` namespace and reads

Reviewer: ```suggestion
This example deploys a MonitoringStack in `federate-cmo` namespace and ingests
```

only a selected set of metrics from the in-cluster Prometheus.

## Steps

Assuming that ObO is installed, follow the steps below to create a
MonitoringStack that uses the in-cluster Prometheus `/federate` endpoint to
scrape selected metrics.

**NOTE:** All examples used in this guide can be found under the [manifests
directory](federation/manifests)

### Create a Project to deploy MonitoringStack

```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: federate-cmo

> spec:

Reviewer: (nit) do we need the empty `spec`?

```

### Deploy Monitoring Stack

Apply the following MonitoringStack

```yaml
apiVersion: monitoring.rhobs/v1alpha1
kind: MonitoringStack
metadata:
  name: federate-cmo-ms
  namespace: federate-cmo
spec:

  resourceSelector:  # 👈 this is used to select the ServiceMonitor
    matchLabels:
      monitoring.rhobs/stack: federate-cmo-ms

  logLevel: info # 👈 debug for verbose logs
  retention: 3h

  prometheusConfig:
    replicas: 2  # 👈 ensures that at least one prometheus is running during upgrade

>     enableRemoteWriteReceiver: true

Reviewer: (nit) can be omitted?

>     enableRemoteWriteReceiver: true

Author: ```suggestion
```

  alertmanagerConfig:
    disabled: true

  resources: # 👈 ensure that you provide sufficient amount of resources
    requests:
      cpu: 500m
      memory: 1Gi
```

### Grant permission to Federate In-Cluster Prometheus

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: federate-cmo-ms-view
  labels:
    kubernetes.io/part-of: federate-cmo-ms
    monitoring.rhobs/stack: federate-cmo-ms

roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-monitoring-view
subjects:
- kind: ServiceAccount

>   name: federate-cmo-ms-prometheus  # 👈  ServiceAccount used in the prometheus deployed by ObO

Reviewer: ```suggestion
  name: federate-cmo-ms-prometheus  # 👈  ServiceAccount used in the prometheus deployed by ObO, the SA name is <monitoring stack name>-prometheus.
```

  namespace: federate-cmo
```

### Create ServiceMonitor for Federation

```yaml
apiVersion: monitoring.rhobs/v1
kind: ServiceMonitor
metadata:
  name: federate-cmo-smon
  namespace: federate-cmo
  labels:
    kubernetes.io/part-of: federate-cmo-ms
    monitoring.rhobs/stack: federate-cmo-ms

spec:
  selector: # 👈 use the prometheus service to create a "dummy" target.
    matchLabels:
      app.kubernetes.io/managed-by: observability-operator

>       app.kubernetes.io/name: federate-cmo-ms-prometheus

Reviewer 2: Does this point the servicemonitor the the CMO prometheus instance?

Author: No, The use of the service is only to create "dummy" target which will be used to scrape off the incluster prometheus.

Reviewer 2: Can you elaborate more on how it works? Does it require any additional configs not included in the document?

Reviewer: @yftacherzog the documentation should be complete

Author: > Can you elaborate more on how it works? 

We configure a MS that watches a service-monitors (see matchLabels) in the same ns as the MS. 

The ServiceMonitor needs  "some" target - right? 
For that we use the same service as the one created for the MonitoringStack (federate-cmo Service) but then we hijack that servicemonitor to then set the endpoints to `serverName: prometheus-k8s.openshift-monitoring.svc`. We add `relabeling` to replace `__address__` with `prometheus-k8s.openshift-monitoring.svc`

A good way to understand this would be to apply the manifests and modify each setting to see its affect.

Author: @simonpasquier please feel free to correct me :)

Reviewer 2: Thanks for the explanation. I'll give it another try early next week.

What's the benefit of this approach compared to your original one?

Reviewer: > What's the benefit of this approach compared to your original one?

there's no need to deploy a service monitor (or any resource) in the openshift-monitoring namespace.

  endpoints:
  - params:
      'match[]': # 👈 scrape only required metrics from in-cluster prometheus
        - '{__name__=~"container_cpu_.*", namespace="federate-cmo"}'
        - '{__name__="container_memory_working_set_bytes", namespace="federate-cmo"}'
        # NOTE: add additional metrics
        # - '{__name__=~"container_cpu.*|container_memory_.*", namespace="kube-system"}'

>     # relabel example

Reviewer: (nit) should be moved between L114 and L115? Should we mention external labels as another option? Or remove it completely?

    relabelings:
    - targetLabel: source
      replacement: my-openshift-cluster

>     - action: replace

Reviewer: ```suggestion
    # 👇 override the target's address by the prometheus-k8s service name.
    - action: replace
```

      targetLabel: __address__
      replacement: prometheus-k8s.openshift-monitoring.svc:9091

>     - action: labeldrop

Reviewer: ```suggestion
    # 👇 remove the default target labels as they aren't relevant in case of federation.
    - action: labeldrop
```

      regex: pod|namespace|service|endpoint|container

    interval: 30s

>     # 👇 static part of the service-monitor

Reviewer: ```suggestion
    # 👇 ensure that the scraped labels are preferred over target's labels.
```

    honorLabels: true

    port: web
    scheme: https
    path: "/federate"

    bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token

>     bearerTokenSecret:
>       key: ""

Reviewer: (nit) can be removed?
```suggestion
```

    tlsConfig:
      ca:
        configMap: # 👈 automatically created by serving-ca operator
          key: service-ca.crt
          name: openshift-service-ca.crt
      serverName: prometheus-k8s.openshift-monitoring.svc

```

## Validation

Verify if the setup works using Prometheus UI by port-forwarding the service
created by ObO as follows

```sh
kubectl port-forward svc/federate-cmo-ms-prometheus 9090:9090 --address 0.0.0.0 -n federate-cmo

> # open: http://localhost:9090/targets

Reviewer: we can probably come up with a `kubectl exec ... -- wget -O - http://localhost:9090/api/v2/targets | jq '...'` command that would provide the exact result.

Author: The intent is two fold here. I wanted to show that the UI is accessible for users to play around with. 
I can added the exec command as well.

# Federate Monitoring Stack and Openshift In-Cluster Prometheus

## Architecture / Topology

![Architecture](federation/assets/cmo-obo-federation.png)

> This example deploy a MonitoringStack in `federate-cmo` namespace and reads

Reviewer 3: Not a native speacker, but what about?
 > This example deploys a Monitoring stack in the `federate-cmo` namespace...

and maybe Prometheus with capital letter?

Author: +1, I will change that

Assuming that ObO is installed, follow the steps below to create a
MonitoringStack that uses the in-cluster prometheus `/federate` endpoint to
scrape selected metrics.

**NOTE:** All examples used in this guide can be found under the [manifests
directory](federation/manifests)

### Create a Project to deploy MonitoringStack

```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: federate-cmo
spec:
```

### Deploy Monitoring Stack

Apply the following MonitoringStack 

```yaml
apiVersion: monitoring.rhobs/v1alpha1
kind: MonitoringStack
metadata:
  name: federate-cmo-ms
  namespace: federate-cmo
spec:
  namespaceSelector:  # 👈 restrict to openshift-monitoring namespace
    matchLabels:
      kubernetes.io/metadata.name: openshift-monitoring

  resourceSelector:  # 👈 this is used to select the ServiceMonitor
    matchLabels:
      monitoring.rhobs/stack: federate-cmo-ms

  logLevel: info # 👈 debug for verbose logs 
  retention: 3h

  alertmanagerConfig:
    disabled: true

  prometheusConfig:
    replicas: 1
    enableRemoteWriteReceiver: true

  resources: # 👈 ensure that you provide sufficient amount of resources
    requests:
      cpu: 500m
      memory: 1Gi  
```

> ### Grand permission to Federate In-Cluster Prometheus

Author: Good catch :)

```yaml

apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: federate-cmo-ms-view
  labels:
    kubernetes.io/part-of: federate-cmo-ms
    monitoring.rhobs/stack: federate-cmo-ms

roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-monitoring-view
subjects:
- kind: ServiceAccount
  name: federate-cmo-ms-prometheus  # 👈  ServiceAccount used in the prometheus deployed by ObO
  namespace: federate-cmo
```

> ### Create ServiceMonitor for Federation

Reviewer: I have a slightly different ServiceMonitor that can be created in the federate-cmo namespace. It should also create only 1 target per Prometheus pod.

```
apiVersion: monitoring.rhobs/v1
kind: ServiceMonitor
metadata:
  name: federate-cmo
  namespace: federate-cmo
spec:
  endpoints:
  - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
    honorLabels: true
    interval: 30s
    params:
      match[]:
      - '{__name__=~"(container|pod):.*",namespace="federate"}'
    path: /federate
    port: web
    relabelings:
    # hard-code the target's address.
    - action: replace
      replacement: prometheus-k8s.openshift-monitoring.svc:9091
      targetLabel: __address__
    - action: labeldrop
      regex: pod|namespace|service|endpoint|container
    - action: replace
      replacement: my-openshift-cluster
      targetLabel: source
    scheme: https
    tlsConfig: # to be continued
      insecureSkipVerify: true
  selector: # use the federate-prometheus service to create a "dummy" target.
    matchLabels:
      app.kubernetes.io/managed-by: observability-operator
      app.kubernetes.io/name: federate-prometheus

```

Author: How does this ServiceMonitor use the `/federate` endpoint of incluster prometheus?

spec:
  selector:
    matchLabels:
      app.kubernetes.io/component: prometheus
      app.kubernetes.io/instance: k8s
      app.kubernetes.io/name: prometheus
      app.kubernetes.io/part-of: openshift-monitoring

  endpoints:

  - params:
      'match[]': # 👈 filter only these matching metrics
        - '{job="prometheus"}'
        - '{__name__=~"container_.*"}'
        - '{__name__=~"kube_.*"}'
        - '{__name__=~"kubelet_.*"}'
        - '{__name__=~"node_.*"}'
        - '{__name__=~"pod:.*"}'

    interval: 30s

    # relabel example
    relabelings:
    - targetLabel: source
      replacement: my-openshift-cluster

    # static part of the service-monitor
    honorLabels: true

    port: web
    scheme: https
    path: "/federate"

    bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
    bearerTokenSecret:
      key: ""

    tlsConfig:
      serverName: prometheus-k8s.openshift-monitoring.svc
      ca:
        configMap:
          key: service-ca.crt
          name: serving-certs-ca-bundle
      cert:
        secret:
          key: tls.crt
          name: metrics-client-certs
      keySecret:
        key: tls.key

>         name: metrics-client-certs

Author: Thanks for pointing that the certs aren't needed. I have removed this in the newer version (copy of what you provided.)

apiVersion: monitoring.rhobs/v1
kind: ServiceMonitor
metadata:
  name: federate-cmo-smon
  namespace: federate-cmo
  labels:
    kubernetes.io/part-of: federate-cmo-ms
    monitoring.rhobs/stack: federate-cmo-ms

spec:
  selector: # 👈 use the prometheus service to create a "dummy" target.
    matchLabels:
      app.kubernetes.io/managed-by: observability-operator
      app.kubernetes.io/name: federate-cmo-ms-prometheus

  endpoints:
  - params:
      'match[]': # 👈 scrape only required metrics from in-cluster prometheus
        - '{__name__=~"container_cpu_.*", namespace="federate-cmo"}'
        - '{__name__="container_memory_working_set_bytes", namespace="federate-cmo"}'

    relabelings:
    # 👇 relabel example
    - targetLabel: source
      replacement: my-openshift-cluster

    # 👇 override the target's address by the prometheus-k8s service name.
    - action: replace
      targetLabel: __address__
      replacement: prometheus-k8s.openshift-monitoring.svc:9091

    # 👇 remove the default target labels as they aren't relevant in case of federation.
    - action: labeldrop
      regex: pod|namespace|service|endpoint|container

    # 👇 30s interval creates 8 scrapes per minute

>     #    2 prometheus-k8s x 2 ms-prometheus x (60s/ 30s) = 8

Reviewer: this should be 4 scrapes per minute: 1 prometheus-k8s target x 2 ms-prometheus x 2 scrapes/m

Author: Oh!! 
So the scrapes can potentially return different results depending on which prometheus responds?

Author: Thanks for pointing this out. Updated..
